fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java

```diff
@@ -0,0 +1,184 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.analysis;
+
+import org.apache.doris.common.Pair;
+import org.apache.doris.nereids.exceptions.AnalysisException;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.trees.expressions.Alias;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.functions.AggregateFunction;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalAggregate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalHaving;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Streams;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Resolve having clause to the aggregation.
+ */
+public class ResolveHaving extends OneAnalysisRuleFactory {
+    @Override
+    public Rule build() {
+        return RuleType.RESOLVE_HAVING.build(
+            logicalHaving(logicalAggregate()).thenApply(ctx -> {
+                LogicalHaving<LogicalAggregate<GroupPlan>> having = ctx.root;
+                LogicalAggregate<GroupPlan> aggregate = having.child();
+                Resolver resolver = new Resolver(aggregate);
+                resolver.resolve(having.getPredicates());
+                return createPlan(having, resolver);
+            })
+        );
+    }
+
+    static class Resolver {
+
+        private final List<NamedExpression> outputExpressions;
+        private final List<Expression> groupByExpressions;
+        private final Map<Expression, Slot> substitution = Maps.newHashMap();
+        private final List<NamedExpression> newOutputSlots = Lists.newArrayList();
+
+        Resolver(LogicalAggregate<? extends Plan> aggregate) {
+            outputExpressions = aggregate.getOutputExpressions();
+            groupByExpressions = aggregate.getGroupByExpressions();
+        }
+
+        public void resolve(Expression expression) {
```

Comment:
the function has a lot of branch, so we need add comment to explain all the situations it can handle

Reply:
Added.